arch/arm/core/aarch32/cpu_idle.S

```diff
@@ -62,19 +62,23 @@ SECTION_FUNC(TEXT, arch_cpu_idle)
 #endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
 #endif /* CONFIG_TRACING */
 
-#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE) \
-	|| defined(CONFIG_ARMV7_R)
-	cpsie i
-#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+#if defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+	/*
+	 * PRIMASK is always cleared on ARMv7-M and ARMv8-M; set PRIMASK to
+	 * prevent interruption before wake-up.
+	 */
+	cpsid i
+
 	/* clear BASEPRI so wfi is awakened by incoming interrupts */
 	eors.n r0, r0
 	msr BASEPRI, r0
-#else
-#error Unknown ARM architecture
-#endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
+#endif /* CONFIG_ARMV7_M_ARMV8_M_MAINLINE */
 
 	wfi
 
+	/* Clear PRIMASK to service any pending interrupt */
+	cpsie i
```

Comment:
In re: #23436, we should `ISB` here to ensure that any pending interrupt (in particular, the one that caused a wake-up) is executed immediately.